libs/wingc/src/jsify.rs

```diff
@@ -41,21 +42,190 @@ const $App = __app(process.env.WING_TARGET);
 const TARGET_APP: &str = "$App";
 
 const INFLIGHT_OBJ_PREFIX: &str = "$Inflight";
+
 pub struct JSifyContext {
 	pub in_json: bool,
 	pub phase: Phase,
 }
 
 pub struct JSifier<'a> {
+	pub diagnostics: Diagnostics,
 	pub out_dir: &'a Path,
 	shim: bool,
 	app_name: String,
 	inflight_counter: RefCell<usize>,
 }
 
+/// Analysizes a resource inflight method and returns a list of fields that are referenced from the
+/// method and which operations are performed on them.
+struct FieldReferenceVisitor<'a> {
+	/// The key is field name, value is a list of operations performed on this field
+	references: BTreeMap<String, IndexSet<String>>,
+
+	/// Used internally by the visitor to keep track of the path to the field
+	path: Vec<String>,
+
+	/// The resource type's symbol env (used to resolve field types)
+	function_def: &'a FunctionDefinition,
+	method_name: &'a Symbol,
+
+	env: SymbolEnvRef,
+	diagnostics: Diagnostics,
+}
+
+impl<'a> FieldReferenceVisitor<'a> {
+	pub fn new(method_name: &'a Symbol, function_def: &'a FunctionDefinition) -> Self {
+		Self {
+			references: BTreeMap::new(),
+			diagnostics: Diagnostics::new(),
+			method_name,
+			path: vec![],
+			function_def,
+			env: function_def.statements.env.borrow().as_ref().unwrap().get_ref(),
+		}
+	}
+
+	pub fn find_refs(mut self) -> (BTreeMap<String, IndexSet<String>>, Diagnostics) {
+		self.visit_scope(&self.function_def.statements);
+		(self.references, self.diagnostics)
+	}
+}
+
+impl<'ast> Visit<'ast> for FieldReferenceVisitor<'_> {
+	fn visit_reference(&mut self, node: &'ast Reference) {
+		match node {
+    	Reference::InstanceMember { object, property } => {
+				match &object.kind {
+        	ExprKind::Reference(r) => {
+						match r {
+							Reference::Identifier(s) => {
+								if s.name == "this" {
+									// resolve the type of the resource by looking up "this" in our environment
+									let resource_type = self.env
+										.lookup(s, None).expect("'this' to be found")
+										.as_variable().expect("'this' to be a variable")
+										.type_;
+
+									// lookup our field in the class environment
+									match &*resource_type {
+            				Type::Resource(r) => {
+											let field_kind = r.env
+												.lookup(property, None).expect("unable to find field in resource env")
+												.as_variable().expect("field is not a variable");
+
+											// we only care about preflight fields (inflight fields are normal references)
+											if field_kind.flight == Phase::Preflight {
```

Comment:
Reorder conditions to bail out early -- `if field_kind.flight != Phase::Preflight` then `break` or `continue` etc. (reduce nesting)

Reply:
Done!